### ISSUE TITLE
Problem with selectFeatures

### ISSUE THREAD

**Issue by massonix** [hemberg-lab/scmap#16]
Dear scmap team,

I am trying to project my scRNA-seq dataset against the Tabula Muris droplet-based reference (which I download from [here](https://github.com/czbiohub/tabula-muris-vignettes/tree/master/data)). This is my code

```{r}
# Load the reference matrix and metadata
metadata_url <- "https://raw.githubusercontent.com/czbiohub/tabula-muris-vignettes/master/data/TM_droplet_metadata.csv"
tm_mat <- readRDS("data/TM_droplet_mat.rds")
tm_metadata <- read_csv(metadata_url, col_names = TRUE)

# Create SingleCellExperiment object
tm_metadata <- as.data.frame(tm_metadata)
rownames(tm_metadata) <- tm_metadata$cell
tm <- SingleCellExperiment(
  assays = list(counts = tm_mat),
  colData = tm_metadata
)

# Index reference dataset
logcounts(tm) <- log((counts(tm) / colSums(counts(tm)) * 10000) + 1)
rowData(tm)$feature_symbol <- rownames(tm)
tm <- selectFeatures(tm, suppress_plot = FALSE)
```
However, I get this error:  Error in rowSums(count == 0) : 
  'x' must be an array of at least two dimensions

I looked for this command in your code but couldn't find it. If you could help me with this it would be very helpful. Thanks in advance for that and for your packages, they are amazing!

Best

Ramon

**Comment by wikiselev** [comment]
Hi, thanks for your message! I've looked into it and it seems that the error occurs here:
https://github.com/hemberg-lab/scmap/blob/d8160c7602770e890e1ca9b8530af9625b6bade4/R/Utils.R#L151
However, if I run the inner scripts manually there is no error, i.e.:
```
count <- as.matrix(counts(tm))
dropouts <- rowSums(count == 0)/cols * 100
```
Not sure what is going now, will need to spend more time on it.

**Comment by massonix** [comment]
Ok, thank you very much!

**Comment by massonix** [comment]
Indeed, if I copy the 3 functions (selectFeatures, linearModel and ggplot_features) from this repo to my code it works fine. However, I still get the same error when running the function loaded by the package...

**Comment by wikiselev** [comment]
I feel the problem is that your count matrix is in sparse format:
```
> counts(tm)[1:3,1:3]
3 x 3 sparse Matrix of class "dgCMatrix"
              10X_P4_0_AAACCTGAGATTACCC 10X_P4_0_AAACCTGAGTGCCAGA
0610005C13Rik                         .                         .
0610007C21Rik                         3                         1
0610007L01Rik                         .                         1
              10X_P4_0_AAACCTGCAAATCCGT
0610005C13Rik                         .
0610007C21Rik                         3
0610007L01Rik                         .
```
Does it work if you just convert it to normal matrix? I haven't sorted out yet how to deal with the sparse formats...

**Comment by yasinkaymaz** [comment]
I also had the same issue and found the discussion here. The errors below seem to disappear when I convert the dgCMatrix to normal matrix:

Error` in rowSums(count == 0) :
  'x' must be an array of at least two dimensions
Calls: selectFeatures -> selectFeatures -> linearModel -> rowSums
Execution halted

Error in rowSums(log_count[dropouts_filter, ]) :
  'x' must be an array of at least two dimensions
Calls: selectFeatures -> selectFeatures -> linearModel -> rowSums
Execution halted


`counts(que_sce) <- as.matrix(counts(que_sce))`

`logcounts(que_sce) <- as.matrix(logcounts(que_sce))`
